docs/design/llvm_sparse_runtime.md

```diff
@@ -0,0 +1,461 @@
+        if (*p_chunk_ptr == nullptr) {
+          auto rt = meta->context->runtime;
+          auto alloc = rt->node_allocators[meta->snode_id];
+          *p_chunk_ptr = alloc->allocate();
+        }
+      });
+    }
+    // 4
+    if (i < chunk_start + chunk_size) {
+      // 4-1
+      *(i32 *)(*p_chunk_ptr + sizeof(Ptr) +
+               (i - chunk_start) * meta->element_size) = data;
+      break;
+    }
+    // 4-2
+    p_chunk_ptr = (Ptr *)(*p_chunk_ptr);
+    chunk_start += chunk_size;
+  }
+  return i;
+}
+```
+
+1. Uses the current length `n` as the index (`i`) to store `data`.
+2. `chunk_strat` tracks the starting index of a given chunk, and always starts at `0` . `p_chunk_ptr` is initialized to the pointer to the first chunk.
+3. Inside the `while` loop, checks if the given chunk slot is empty first, and allocates a new chunk if so.
+4. Compares if the determined index `i` falls within the current chunk.
+   1. If so, stores `data` into the corresponding slot in this chunk. Note that the first `sizeof(Ptr)` bytes are skiped: they are reserved to store the address of the next chunk.
+   2. Otherwise, jumps to the next chunk.
+
+# Runtime
+
+Runtime for the LLVM backends is in https://github.com/taichi-dev/taichi/blob/master/taichi/runtime/llvm/runtime.cpp. Note that this file is *NOT* linked into Taichi's core C++ library. Instead, it is compiled into a LLVM byte code file (`.bc`). Upon starting Taichi, the `.bc` file is loaded back into the memory, de-serialized into an `llvm::Module`, and linked together with the JIT compiled Taichi kernels. This design has the advantage that the runtime code can be written once and shared between backends using LLVM, such as CPU and CUDA. In addition, the sparse runtime can be implemented in a language with enough abstraction (i.e., C++ instead of raw LLVM IR).
+
+The core data structure of this runtime is [`LLVMRuntime`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L543), which holds a handful of data:
+
+* All the root SNodes info (`roots` and `root_mem_sizes`)
+* SNode memory allocators
+* Random states for supporting `ti.random` (`rand_states`)
+* Print and error message buffers
+* ...
+
+We will explain how the SNode memory allocator is implemented, which is the bedrock of the sparse SNodes.
+
+## `NodeManager`: a recycling memory allocator
+
+Each SNode is associated with its own memory allocator. These allocators are stored in an array, [`node_allocators`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L562).
+
+The allocator is of type [`NodeManager`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L619). It contains [three linked lists](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L627):
```

Comment:
The ListManager section below seems to come out of no where :P Shall wel replace ` three linked lists` with `three linked list like ListManagers`? ;)

Reply:
Yup, done